From: Dan Colish <colish@chromium.org>
Date: Fri, 20 Nov 2015 10:48:31 -0800

dnsmasq: OPT_NO_FORK was not honored for tcp requests

When handling tcp dns queries the OPT_NO_FORK option was not being
checked. This meant the daemon was forking children unexpectedly.

diff --git a/src/dnsmasq.c b/src/dnsmasq.c
index f4a89fc..dbf7de5 100644
--- a/src/dnsmasq.c
+++ b/src/dnsmasq.c
@@ -1558,7 +1558,9 @@ static void check_dns_listeners(fd_set *set, time_t now)
 	      close(confd);
 	    }
 #ifndef NO_FORK
-	  else if (!option_bool(OPT_DEBUG) && (p = fork()) != 0)
+	  else if (!option_bool(OPT_DEBUG) &&
+                   !option_bool(OPT_NO_FORK) &&
+                   (p = fork()) != 0)
 	    {
 	      if (p != -1)
 		{
@@ -1595,7 +1597,7 @@ static void check_dns_listeners(fd_set *set, time_t now)
 #ifndef NO_FORK
 	      /* Arrange for SIGALARM after CHILD_LIFETIME seconds to
 		 terminate the process. */
-	      if (!option_bool(OPT_DEBUG))
+	      if (!option_bool(OPT_DEBUG) && !option_bool(OPT_NO_FORK))
 		alarm(CHILD_LIFETIME);
 #endif
 
@@ -1624,7 +1626,7 @@ static void check_dns_listeners(fd_set *set, time_t now)
 		    close(s->tcpfd);
 		  }
 #ifndef NO_FORK		   
-	      if (!option_bool(OPT_DEBUG))
+	      if (!option_bool(OPT_DEBUG) && !option_bool(OPT_NO_FORK))
 		{
 		  flush_log();
 		  _exit(0);
-- 
2.6.0.rc2.230.g3dd15c0

